src/Middleware/HttpLogging/src/W3CLogger.cs

```diff
@@ -0,0 +1,181 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using Microsoft.Extensions.Options;
+using System.Collections.Generic;
+using Microsoft.AspNetCore.Http;
+using System.Numerics;
+using System.Text;
+using Microsoft.Net.Http.Headers;
+using System.Text.RegularExpressions;
+using System.Globalization;
+using System.Diagnostics;
+
+namespace Microsoft.AspNetCore.HttpLogging
+{
+    internal class W3CLogger : IDisposable
+    {
+        private readonly W3CLoggerProcessor _messageQueue;
+        private readonly IOptionsMonitor<W3CLoggerOptions> _options;
+        // Subtract 8 to account for flags that represent groups (e.g. "RequestHeaders", "All", "None")
+        private readonly int _fieldsLength = Enum.GetValues(typeof(W3CLoggingFields)).Length - 8;
+
+        internal W3CLogger(IOptionsMonitor<W3CLoggerOptions> options)
+        {
+            _options = options;
+            _messageQueue = InitializeMessageQueue(_options);
+        }
+
+        // Virtual for testing
+        internal virtual W3CLoggerProcessor InitializeMessageQueue(IOptionsMonitor<W3CLoggerOptions> options)
+        {
+            return new W3CLoggerProcessor(options);
+        }
+
+        public void Dispose() => _messageQueue.Dispose();
+
+        public void Log(IReadOnlyCollection<KeyValuePair<string, string?>> state)
```

Comment:
Since this is an internal interface, why not just force the middleware to create the `elements` array in `var elements = new string[_fieldsLength];`? That avoids the list allocation and the switch on the string which isn't a jump table. We could use ArrayPool<string> if we're careful to clear the references too.